nf_core/modules.py

```diff
@@ -231,3 +236,157 @@ def has_valid_pipeline(self):
         if not os.path.exists(main_nf) and not os.path.exists(nf_config):
             log.error("Could not find a main.nf or nextfow.config file in: {}".format(self.pipeline_dir))
             return False
+
+
+class ModulesTestHelper(object):
+    def __init__(
+        self,
+        module_name,
+        test_name=None,
+        test_command=None,
+        test_tags=[],
+        test_input_results_dir=None,
+        run_test=False,
+        test_yml_output_path=None,
+    ):
+        self.module_name = module_name
+        self.test_input_results_dir = test_input_results_dir
+        self.run_test = run_test
+        self.test_yml_output_path = test_yml_output_path
+        self.modules_dir = os.path.join("software", *module_name.split("/"))
+        self.test_yaml = {
+            "name": test_name,
+            "command": test_command,
+            "tags": test_tags,
+            "files": [],
+        }
+
+    # Add custom dumper class to prevent overwriting the global state
+    # This prevents yaml from changing the output order
+    # See https://stackoverflow.com/a/52621703/1497385
+    class CustomDumper(yaml.Dumper):
+        def represent_dict_preserve_order(self, data):
+            return self.represent_dict(data.items())
+
+    CustomDumper.add_representer(dict, CustomDumper.represent_dict_preserve_order)
+
+    def check_inputs(self):
+        """ Do more complex checks about supplied flags. """
+        # First, sanity check that the module directory exists
+        if not os.path.isdir(self.modules_dir):
+            raise UserWarning(f"Cannot find directory '{self.modules_dir}'! Should be TOOL/SUBTOOL or TOOL")
+
+        # Sanity check + assign test run flags
+        if self.run_test and self.test_input_results_dir is not None:
+            raise UserWarning(f"Either supply '--input' or '--run-test', not both")
+        if not self.run_test and self.test_input_results_dir is None:
+            raise UserWarning(
+                f"Either supply a test output directory with '--input' or trigger a run with '--run-test'"
+            )
+
+        # Check that the output YAML file does not already exist
+        # TODO: Instead of clobbering can parse + append to list if already there
+        if self.test_yml_output_path is not None and os.path.exists(self.test_yml_output_path):
+            raise UserWarning(f"Test YAML file already exists! '{self.test_yml_output_path}'")
```

Comment:
Yes, maybe not copying that file is the best approach.. Though to be honest if this tool just overwrites files then it's not a terrible idea to have it there I guess...? Can't decide whether it's better to have a dummy file that will be overwritten so that people know that there should be a file there, or not to have it to force people to figure out that this command exists 🤔

Reply:
I guess it's fine to have it there with a big `// TODO` saying to run this command?